### PR TITLE
mgr: remove non-existent MDS daemons from FSMap

### PR DESCRIPTION
**Pull request by SpandanKumarSahu**
Fixes: http://tracker.ceph.com/issues/17453
Signed-off-by: Spandan Kumar Sahu <spandankumarsahu@gmail.com>

### CODE REVIEW

**Comment by SpandanKumarSahu** [comment]
The build is successful on my machine. However the Jenkin checks suggest otherwise. Please let me know, where I am going wrong.

File: src/mgr/Mgr.cc [hunk 1]
```diff
@@ -521,6 +523,13 @@ void Mgr::handle_fs_map(MFSMap* m)
   for (const auto &i : mds_info) {
     const auto &info = i.second;
 
+    if(!new_fsmap.gid_exists(i.first)){
```
Comment (liewegas) [inline src/mgr/Mgr.cc]: "if ("

(missing a space)

**Comment by liewegas** [comment]
Otherwise this looks good!

File: src/mgr/Mgr.cc [hunk 2]
```diff
@@ -521,6 +523,13 @@ void Mgr::handle_fs_map(MFSMap* m)
...
+      continue;
+    }
+
+    //Remember which MDS exists so that we can cull any that don't
```
Comment (tchaikov) [inline src/mgr/Mgr.cc]: please add a space after `//`

**Comment by SpandanKumarSahu** [comment]
Made the necessary changes.
@liewegas @tchaikov

File: src/mgr/Mgr.cc [hunk 3]
```diff
@@ -521,6 +523,13 @@ void Mgr::handle_fs_map(MFSMap* m)
   for (const auto &i : mds_info) {
     const auto &info = i.second;
 
+    if (!new_fsmap.gid_exists(i.first)){
+      continue;
+    }
+
+    // Remember which MDS exists so that we can cull any that don't
+    names_exist.insert(stringify(info.name));
```
Comment (tchaikov) [inline src/mgr/Mgr.cc]: `info.name` is a string already, i don't think we need to `stringify()` it.

**Comment by SpandanKumarSahu** [comment]
@tchaikov

**Comment by liewegas** [comment]
causes 

2017-05-06 06:11:05.391477 7f26da7b0700  0 log_channel(cluster) log [INF] : HEALTH_WARN; insufficient standby daemons available: have 0; want 1 more

/a/sage-2017-05-06_05:54:44-rados-wip-sage-testing2---basic-smithi/1108765

**Comment by SpandanKumarSahu** [comment]
@liewegas Why is that so? Apparently, the code should remove only the ones which don't exist, right?

**Comment by tchaikov** [comment]
i think this issue is unrelated to this change. instead, it is stemmed from `tasks/cephfs/test_failover.py: test_grow_shrink()`. in this test, we calls
```py
        fs_a.set_max_mds(3)
```
which set the `max_mds` to 3. 

so all up mds were active after that
```
2017-05-06 06:11:04.602396 7f26da7b0700 10 mon.a@0(leader).mds e11 e11: 3/3/3 up {0=a=up:active,1=c=up:active,2=b=up:active}
```

in other words, none of the 3 MDS is standby anymore. but we set the standby count to 1 in `MDSMap::check_health()`. @batrick is this expected?

**Comment by batrick** [comment]
Yes, that's expected. I will make a PR for test_grow_shrink that turns off that warning.

**Comment by tchaikov** [comment]
@batrick thank you!

**Comment by tchaikov** [comment]
@liewegas as @batrick has confirmed this, can we merge this change?

**Comment by ukernel** [comment]
agree with tchaikov. the healthe warning is unrelated

**Comment by liewegas** [comment]
thanks everyone!

**Comment by batrick** [comment]
FYI: https://github.com/ceph/ceph/pull/15035

**Pull request merged**